keel-bakery/src/main/kotlin/com/netflix/spinnaker/keel/bakery/resource/ImageHandler.kt

```diff
@@ -0,0 +1,135 @@
+import com.netflix.spinnaker.keel.model.OrchestrationRequest
+import com.netflix.spinnaker.keel.model.OrchestrationTrigger
+import com.netflix.spinnaker.keel.orca.OrcaService
+import com.netflix.spinnaker.keel.persistence.ArtifactRepository
+import com.netflix.spinnaker.keel.plugin.ResolvableResourceHandler
+import com.netflix.spinnaker.keel.plugin.ResourceDiff
+import com.netflix.spinnaker.keel.plugin.ResourceNormalizer
+import kotlinx.coroutines.runBlocking
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+class ImageHandler(
+  override val objectMapper: ObjectMapper,
+  override val normalizers: List<ResourceNormalizer<*>>,
+  private val cloudDriver: CloudDriverService,
+  private val artifactRepository: ArtifactRepository,
+  private val baseImageCache: BaseImageCache,
+  private val orcaService: OrcaService,
+  private val igorService: ArtifactService
+) : ResolvableResourceHandler<ImageSpec, Image> {
+
+  override val apiVersion = SPINNAKER_API_V1.subApi("bakery")
+  override val supportedKind = ResourceKind(
+    apiVersion.group,
+    "image",
+    "images"
+  ) to ImageSpec::class.java
+
+  override fun generateName(spec: ImageSpec): ResourceName =
+    ResourceName("bakery:image:${spec.artifactName}")
+
+  override fun desired(resource: Resource<ImageSpec>): Image =
+    with(resource.spec) {
+      val latestVersion = artifactRepository
+        .versions(DeliveryArtifact(artifactName, DEB))
+        .first()
+      Image(
+        baseAmiVersion = baseImageCache.getBaseImage(baseOs, baseLabel),
+        // TODO: there must be a canonical way of composing this
+        appVersion = latestVersion,
+        regions = regions
+      )
+    }
+
+  override fun current(resource: Resource<ImageSpec>): Image? =
+    runBlocking {
+      cloudDriver.images("aws", resource.spec.artifactName)
+        .await()
+        .firstOrNull()
+        ?.let {
+          // TODO: this is pretty crude to say the least
+          val tags = it.tagsByImageId.values.first()!!
+          Image(
+            tags.getValue("base_ami_version")!!,
+            tags.getValue("appversion")!!,
+            it.amis.keys
+          )
+        }
+    }
+
+  override fun upsert(
+    resource: Resource<ImageSpec>,
+    resourceDiff: ResourceDiff<Image>
+  ): List<TaskRef> {
+    val taskRef = runBlocking {
+      val (_, application, version) = resourceDiff.desired.appVersion.let { appVersion ->
+        Regex("([\\w_]+)-(.+?)/.*")
+          .find(appVersion)
+          ?.groupValues
+          ?: throw IllegalStateException("Could not parse app version $appVersion")
+      }
+      val artifact = igorService.getArtifact(application, version)
+      log.info("baking new image for {}", resource.spec.artifactName)
+      orcaService.orchestrate(
+        OrchestrationRequest(
+          name = "Bake image for ${resource.spec.artifactName}",
+          application = resource.spec.artifactName, // TODO: is this always the same? Does it matter?
```

Comment:
I don't know if we want to use artifact name here - the name most likely won't correspond to an app (think of things like redis that are common or packages that have a different name from their app, like if our package was `keel-nflx`). Either, we should have someone define an app in their config to run tasks in, or we should just run it in keel for now.

An alternative for finding the app for an image (for netflix) is to build an endpoint in swabbie. Swabbie already has a cache of these things and is pulling internal data to see if an image is used anywhere in any app. We could easily expose this to get that application. This doesn't solve the "many apps use one package" case though.

Reply:
Right now we don't have a general way of tying resources to applications. I'll make it bake as Keel for now & we can revisit when we decide to tackle that issue.